server/src/main/java/org/opensearch/search/aggregations/bucket/terms/MultiTermsAggregator.java

```diff
@@ -414,6 +372,74 @@ public void close() {
         }
     }
 
+    static class MultiValuesSourceCollectorImpl implements MultiTermsValuesSourceCollector {
+
+        private final List<InternalValuesSourceCollector> collectors;
+        private final BytesStreamOutput scratch;
+        private final BytesKeyedBucketOrds bucketOrds;
+        private final BucketsAggregator aggregator;
+        private final LeafBucketCollector sub;
+
+        private final boolean collectViaAggregator;
+
+        public MultiValuesSourceCollectorImpl(
+            List<InternalValuesSourceCollector> collectors,
+            BytesStreamOutput scratch,
+            BytesKeyedBucketOrds bucketOrds,
+            BucketsAggregator aggregator,
+            LeafBucketCollector sub
+        ) {
+            this.collectors = collectors;
+            this.scratch = scratch;
+            this.bucketOrds = bucketOrds;
+            this.aggregator = aggregator;
+            this.sub = sub;
+            this.collectViaAggregator = aggregator != null && sub != null;
+        }
+
+        @Override
+        public void apply(int doc, long owningBucketOrd) throws IOException {
+            List<List<TermValue<?>>> collectedValues = new ArrayList<>();
+            for (InternalValuesSourceCollector collector : collectors) {
+                collectedValues.add(collector.apply(doc));
+            }
+            scratch.seek(0);
+            scratch.writeVInt(collectors.size()); // number of fields per composite key
+            cartesianProductRecursive(collectedValues, 0, owningBucketOrd, doc);
+        }
+
+        /**
+         * Cartesian product using depth first search.
+         */
+        private void cartesianProductRecursive(List<List<TermValue<?>>> collectedValues, int index, long owningBucketOrd, int doc)
+            throws IOException {
+            if (collectedValues.size() == index) {
+                // Avoid performing a deep copy of the composite key
+                long bucketOrd = bucketOrds.add(owningBucketOrd, scratch.bytes().toBytesRef());
+                if (collectViaAggregator) {
+                    if (bucketOrd < 0) {
+                        bucketOrd = -1 - bucketOrd;
+                        aggregator.collectExistingBucket(sub, doc, bucketOrd);
+                    } else {
+                        aggregator.collectBucket(sub, doc, bucketOrd);
+                    }
+                }
+                return;
+            }
+
+            long position = scratch.position();
+            List<TermValue<?>> values = collectedValues.get(index);
+            int numIterations = values.size();
+            for (int i = 0; i < numIterations; i++) {
+                TermValue<?> value = values.get(i);
+                value.writeTo(scratch); // encode the value
+                cartesianProductRecursive(collectedValues, index + 1, owningBucketOrd, doc); // dfs
```

Comment:
Wondering if we should have a TODO or a further action item on seeing if we can make this recursive call to iterative if that gives a performance improvement? 
This can be a different experimentation itself outside the scope of this PR. Do you think its worth checking?

Reply:
Yes I had seen some improvements with making the method iterative. Had tried 2 variations one being simple nested for loops ( only for 2 fields in aggregation ) and another one was generic for any number of terms in aggregation. The former was faster than the latter due to use of division and modulus in the latter. 

Will create a separate PR for that as there are some other things I want to try out.